src/dnmd/write.c

```diff
@@ -0,0 +1,957 @@
+
+    mdcursor_t prior_row = row;
+    if (md_cursor_move(&prior_row, -1) && !CursorNull(&prior_row))
+    {
+        if (!is_row_sorted_with_next_row(keys, count_keys, table->table_id, prior_row, row))
+            return false;
+    }
+    mdcursor_t next_row = row;
+    if (!md_cursor_next(&next_row) && !CursorEnd(&next_row))
+    {
+        if (!is_row_sorted_with_next_row(keys, count_keys, table->table_id, row, next_row))
+            return false;
+    }
+
+    return true;
+}
+
+void md_commit_row_add(mdcursor_t row)
+{
+    mdtable_t* table = CursorTable(&row);
+    assert(table->is_adding_new_row);
+
+    // If the table was previously sorted,
+    // validate that the current row is sorted with respect to the prior and following rows.
+    if (table->is_sorted)
+    {
+        table->is_sorted = validate_row_sorted_within_table(row);
+    }
+
+
+    table->is_adding_new_row = false;
+}
+
+bool sort_list_by_column(mdcursor_t parent, col_index_t list_col, col_index_t col)
+{
+    mdcursor_t range;
+    uint32_t count;
+    bool success = md_get_column_value_as_range(parent, list_col, &range, &count);
+    assert(success);
+    (void)success;
+
+    // A one element range is always sorted.
+    if (count == 1)
+        return true;
+
+    void* cursor_order_buffer = malloc((sizeof(mdcursor_t) + sizeof(int32_t)) * count);
+    mdcursor_t* correct_cursor_order = cursor_order_buffer;
+    int32_t* correct_cursor_order_ids = (int32_t*)(((mdcursor_t*)cursor_order_buffer) + count);
+
+    if (correct_cursor_order == NULL)
+        return false;
+
+    memset(correct_cursor_order, 0, sizeof(*correct_cursor_order) * count);
+
+    bool need_to_update = false;
+    mdcursor_t list_item = range;
+    int32_t next_index = 0;
+    // Gather cursors to all of the rows in the list,
+    // and put them in the order specified by the column.
+    for (uint32_t i = 0; i < count; i++, md_cursor_next(&list_item))
+    {
+        mdcursor_t target;
+        if (!md_resolve_indirect_cursor(list_item, &target))
+            return false;
+
+        uint32_t sequence_number;
+        if (1 != md_get_column_value_as_constant(target, col, 1, &sequence_number))
+        {
+            free(cursor_order_buffer);
+            assert(!"Failed to read constant column from target cursor");
+            return false;
+        }
+
+        assert(CursorNull(&correct_cursor_order[next_index]));
+        correct_cursor_order[next_index] = target;
+        correct_cursor_order_ids[next_index] = sequence_number;
+
+        // Sequence ids need to be in ascending order to be sorted.
+        if (next_index > 0
+            && correct_cursor_order_ids[next_index - 1] > correct_cursor_order_ids[next_index])
+        {
+            // Do a simple insertion sort as we go.
+            // It's unlikely we'll need to sort,
+            // and even when we do, we'll likely be mostly sorted anyway.
```

Comment:
Anything to assert this assumption?

Reply:
This function would only do work if a parameter is added in the middle of a parameter list (or an attribute is added on a parameter in the middle of the signature). In this scenario, the list would be sorted for all but the newest element. Theoretically we could also get here for a delta that adds a new method, but all delta producers today emit the parameter entries in order when a new method is added.

A tool that manually constructs a delta that has the parameters out of order is possible, but it would have to be specifically written to do so.